### PR TITLE
[h2] Acknowledge window updates aggressively

### LINKED ISSUE #1013: flow control may starve gRPC streams
Consider the situation when an endpoint sends two gRPC messages (_msgA_ and _msgB_) on a gRPC stream (with max frame length, _F_, and an initial window size, _W=2F_). _msgA_ has a length of F-1, and _msgB_ has a length of F+2.
```
        +-------------+
frame0: |aaaaaaaaaaa|b|
        +-------------+
frame1: |bbbbbbbbbbbbb|
        +-------------+
frame2: |b|           |
        +-------------+
```

The receiver of this gRPC stream _should_ operate as follows:
1. Read _msgA_
2. Release _msgA_
    1. Send `WINDOW_UPDATE ${F-1}`
2. Read _msgB_

However, this doesn't work as designed!  Netty wont actually send the `WINDOW_UPDATE` with F-1 bytes because Netty's default flow controller only sends window updates of at least 0.5*W bytes.  But we can't fully read _msgB_ until the sender transmits _frame2_, and the sender may not transmit _frame2_ until the window is updated by at least 1 byte.  Because of all of this, streams may hang and stop transmitting updates.

While this problem is clearly manifested in our gRPC implementation, I think this issue _could theoretically_ cause problems in normal H2 routing, depending on the HTTP/2 implementations of peers.

I'd like to look at replacing Netty's default flow controller for linkerd.

### PR DESCRIPTION
**Pull request by olix0r**
Problem

By default, netty does not send a WINDOW_UPDATE until the available window is at ≥50% capacity. This can lead to starvation (especially for gRPC).

Solution

Change the default window increment ratio from 0.5 to 0.95 so that WINDOW_UPDATE frames are sent much more aggressively.

This branch has been tested against the ver/grpc-interop-server branch (#1024, to be merged soon after this), confirming that this fixes #1013.

In a follow-up, these parameters will be made configurable.

### CODE REVIEW

File: finagle/h2/src/main/scala/io/netty/handler/codec/http2/H2FrameCodec.scala [hunk 1]
```diff
@@ -168,21 +167,50 @@ class H2FrameCodec(
 
 object H2FrameCodec {
 
-  private lazy val frameLogger = new Http2FrameLogger(io.netty.handler.logging.LogLevel.DEBUG, getClass)
+  /** Aggressively acknowledge window updates */
+  val DefaultWindowUpdateRatio = 0.95f
```
Comment (stevej) [inline finagle/h2/src/main/scala/io/netty/handler/codec/http2/H2FrameCodec.scala]: I have the same question here about why it's not 1.0f

File: finagle/h2/src/main/scala/io/netty/handler/codec/http2/H2FrameCodec.scala [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+
+      case msg => ctx.write(msg, promise); ()
+    }
+
+}
+
+object H2FrameCodec {
+
+  /** Aggressively acknowledge window updates */
+  val DefaultWindowUpdateRatio = 1.0f
```
Comment (adleong) [inline finagle/h2/src/main/scala/io/netty/handler/codec/http2/H2FrameCodec.scala]: To make sure I understand: this means that any frames that are released will immediately trigger a window update?  Whereas a ratio of 0.95f would mean that releasing frames will not trigger a window update until 5% of the window size has been released?

Comment (olix0r) [inline finagle/h2/src/main/scala/io/netty/handler/codec/http2/H2FrameCodec.scala]: correct

**Comment by olix0r** [comment]
It turns out that ratios have to be on `(0.0, 1.0)` (exclusive).

**Comment by olix0r** [comment]
⭐️ myself

**Pull request closed without merging**